### DISCUSSION TITLE
Questions about parity, higher order irreps, and others in e3nn

### DISCUSSION THREAD

**Discussion by Karbo123** [e3nn/e3nn#287] [Q&A]
Thanks for releasing this great work for equivariantly learning from 3D data in group E(3)! I am recently interested in reconstructing 3D surface from oriented point cloud input, and deeply believe that equivariant learning should be a promising research direction in this field, because equivariant network naturally captures the symmetries of rotations and reflections in the input point cloud and decouples the orientations from oriented inputs.

After roughly reading some papers, I have some basic questions about the concepts of parity and irreps used in e3nn.
1. **What is the first paper to propose incorporating parity in the theory of e3nn?** As far as I know, e3nn is probably based on the theory proposed in tensor field network (TFN). But TFN only learns on group SE(3), so how does TFN classify the two chiral shapes in experiment of tetris?
2. It is not obvious for me that **how to use e3nn to only learn from group SE(3)?** Is it okay to just set the parity of order-l harmonic to even if order l is an even number, and to odd if order l is an odd number, so as to only learn from group SE(3)?
3. **Any suggestion to use higher order irreps in the network?** As for me, the inputs and outputs are generally scalars(l==0) and vectors(l==1). What is the benefit to use higher order irreps in the hidden layers? Will it increase the learning ability of the network?
4. **Any example of expressing a meaningful tensor with higher order irreps (l>=2)?** Should the curvature of a point on the surface be expressed using order-2 irrep? Any intuition for the example?
5. From my experience of doing some experiments, learning from the equivariant network built from point convolutions (i.e. graph convolution using tensor products between harmonics of edge vectors and irreps of nodes) seems to be hard to learn a global meaning (e.g. a global rotation of a point cloud)? **Any suggestion for improving the ability of learning a global meaning?**
6. Besides, why is TFN designed to learn from tensor products between harmonics of edge vectors and irreps of nodes? I think that such design may only focus on low-level geometric structure as the point convolution always need to interact with the low-level edge irreps instead of other high-level semantic irreps. Can we design the network in another way? (e.g. tensor products between two irreps taken from two nodes, or just use interleaved linear layers and equivariant non-linearities.)

I'm sorry if there are some grammar mistakes. It seems like a lot of questions, but I will be appreciated if any suggestion is provided.

**Comment by mariogeiger** [comment]
Hi, these are interesting question. I will answer to the questions I know the answer.

1. As far as I know there is no paper proposing it, we are currently writing it.
2. If you declare all the irreps as even (the irreps of spherical harmonics becomes `0e+1e+2e...`) your network becomes equivariant to SE(3) instead of E(3). If you are equivariant to SE(3) the two chiral shapes are two different objects related by no symmetries.
3. @simonbatzner on his experiments with his tool https://github.com/mir-group/nequip observed the user of higher l to be beneficial.
4. These quantities are often expressed as cartesian tensors (tensors with multiple indices) like for instance the stress tensor or the elasticity tensor. You can use [ReducedTensorProducts](https://docs.e3nn.org/en/stable/api/o3/o3_reduce.html) to obtain the change of basis into irreps.
5. I don't have enough experience to know, @blondegeek any idea?
6. Yes you can do that, you should try for sure!

**Reply by Karbo123** [reply]
Thanks, the replies on questions 1~4 address my questions well.

**Comment by blondegeek** [comment]
Hi Karbo,

Thanks for the questions.

(5) As far as learning a global meaning, there's no one answer and will
definitely depend on the specifics of your application. Regarding learning
the rotation of a point cloud, I want to clarify that only relative
rotations are well defined -- e.g. between which two coordinate systems you
are defining the rotation. You can do a variety of pooling procedures to
predict a global 3x3 matrix, which you may choose to use to represent a 3D
rotation, from per point features. You could do this pooling all at once or
at different resolutions. A example of a very different way to get global
information -- in material science, we tend to use Fourier / reciprocal
space to describe long-range or "global properties" -- this may make less
sense for your case because you're not dealing with periodic objects -- but
just wanted to give you an example about how what "global" properties mean
can vary a lot based on the application.

(6) There are so many equivariant interactions that can be used that have
not been sufficiently explored! With e3nn, we've provided a library that
has all the modular pieces you need to build up operations of arbitrary
complexity. You should definitely feel free to test out operations beyond
any of the architectures specified in tensor field networks, 3d steerable
cnns, cormorant, equivariant transformers, etc. Let us know if you invent
an interaction that you find particularly expressive :)


Hope that helps a bit! Please feel free to reach out again with more
questions.
Tess

**Reply by Karbo123** [reply]
Thanks for the constructive comments! After reading these, I have the following thoughts.
(5.1) I believe relative rotations are definitely not a problem. As for the tasks of rotation estimation, the canonical (standard pose) point clouds are always available, the network can be designed to "memorize" the canonical rotation matrix (although it is an identity 3x3 matrix at training time) via learning from canonical point cloud, and equivariance ensures that it can generalize to rotated poses as well, which is roughly verified in my experiments.
(5.2) As for me, the biggest problem might be that the training data of point clouds contains too many points (for a toy object > 1000 points), and existing designed networks are not very suitable for this situation, since the receptive field of the network is relative small (but increasing the radius consumes too much CUDA memory), so that they cannot perceive the **whole** (global meaning) point cloud. Besides, the learned irreps of the network are very sensitive to both the sampling density and the point distribution (although sampled from the same object surface), this largely limits the learning power of the equivariant networks. Nevertheless, some recently proposed works on equivariant point cloud processing implies that the theory of e3nn might be the most general one, and I believe e3nn can perform better than others.
(6) I think there should be a more appropriate and efficient way to process point clouds with low memory consumption and high computational efficiency. I could try more.
